src/python/pants/task/simple_codegen_task.py

```diff
@@ -293,6 +336,20 @@ def _inject_synthetic_target(
 
     return synthetic_target
 
+  def _original_export_specs(self, target):
+    collected_original_exports = getattr(target, 'export_specs', None)
+    if collected_original_exports is None:
+      collected_original_exports = list()
+
+    # resolve and maybe synthetic-ify original exports
```

Comment:
Yeah. That comment is not very clear.
What I'm trying to do here is to cover these three cases
* exported codegen targets with no synthetic derivative yet
* exported non-codegen targets
* exported codegen targets with a synthetic derivative
In the first 2 cases, just the target is added. In the last, both the target and the synthetic derivative is added.

After thinking about it, this is is wrong.

Hm. I think if I switch to topo order, within a codegen task, the synthetic targets will be injected correctly. Thinking about it a bit more deeply, there's still an issue though which is that other codegen tasks' targets could be exported. So, in order to move the synthetic target detection from Target.exports like I'm doing, I'd need to be able to inject exports after the fact to already constructed targets. 😞 

I'm now wondering whether I should break this up into
1. a change that adds the extra exports and copies the original exports without any special handling synthetic targets. I'd revert the changes to Target and simplify things here.
2. a change that ensures that synthetic targets replace their codegen targets in exports during codegen

Doing 2) right now is tricky because doing it properly requires modifications. This change as it is now, tries to do that, but it'll introduce weird edgecases if codegen targets from different tasks interact.

Reply:
> Thinking about it a bit more deeply, there's still an issue though which is that other codegen tasks' targets could be exported.

In theory maybe, but not in practice.